File: genium/src/main/java/com/here/genium/generator/cbridge/CBridgeTypeMapper.java
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.cbridge;

import static com.here.genium.generator.cbridge.CBridgeNameRules.BASE_HANDLE_IMPL_FILE;
import static com.here.genium.generator.cbridge.CBridgeNameRules.BASE_REF_NAME;
import static com.here.genium.generator.cbridge.CppTypeInfo.TypeCategory.BUILTIN_BYTEBUFFER;
import static com.here.genium.generator.cbridge.CppTypeInfo.TypeCategory.CLASS;
import static com.here.genium.generator.cbridge.CppTypeInfo.TypeCategory.ENUM;
import static com.here.genium.generator.cbridge.CppTypeInfo.TypeCategory.STRUCT;
import static com.here.genium.model.cbridge.CType.VOID;
import static com.here.genium.model.common.InstanceRules.isInstanceId;

import com.here.genium.common.FrancaTypeHelper;
import com.here.genium.generator.cpp.CppLibraryIncludes;
import com.here.genium.generator.cpp.CppNameResolver;
import com.here.genium.model.cbridge.CBridgeIncludeResolver;
import com.here.genium.model.cbridge.CType;
import com.here.genium.model.common.Include;
import com.here.genium.model.cpp.CppIncludeResolver;
import com.here.genium.model.cpp.CppTypeRef;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import org.franca.core.franca.FArrayType;
import org.franca.core.franca.FBasicTypeId;
import org.franca.core.franca.FEnumerationType;
import org.franca.core.franca.FInterface;
import org.franca.core.franca.FMapType;
import org.franca.core.franca.FModelElement;
import org.franca.core.franca.FStructType;
import org.franca.core.franca.FType;
import org.franca.core.franca.FTypeDef;
import org.franca.core.franca.FTypeRef;

public class CBridgeTypeMapper {

  private static final Include BASE_HANDLE_IMPL_INCLUDE =
      Include.Companion.createInternalInclude(CBridgeNameRules.BASE_HANDLE_IMPL_FILE);

  private final CppIncludeResolver cppIncludeResolver;
  private final CppNameResolver cppNameResolver;
  private final CBridgeIncludeResolver includeResolver;
  private final String enumHashType;
  private final CppTypeInfo byteBufferTypeInfo;

  public CBridgeTypeMapper(
      final CppIncludeResolver cppIncludeResolver,
      final CppNameResolver cppNameResolver,
      final CBridgeIncludeResolver includeResolver,
      final String enumHashType,
      final String byteBufferType) {
    this.cppIncludeResolver = cppIncludeResolver;
    this.cppNameResolver = cppNameResolver;
    this.includeResolver = includeResolver;
    this.enumHashType = enumHashType;
    this.byteBufferTypeInfo =
        new CppTypeInfo(
            byteBufferType,
            new CType(BASE_REF_NAME),
            CType.BYTE_ARRAY_REF,
            BUILTIN_BYTEBUFFER,
            Collections.singletonList(
                Include.Companion.createInternalInclude(BASE_HANDLE_IMPL_FILE)));
  }

  public CppTypeInfo mapType(final FTypeRef typeRef) {
    FType derived = typeRef.getDerived();
    CppTypeInfo typeResult;
    if (derived != null) {
      typeResult = mapType(derived);
    } else {
      typeResult = mapPredefined(typeRef);
    }

    if (FrancaTypeHelper.isImplicitArray(typeRef)) {
      typeResult = CArrayMapper.createArrayReference(typeResult);
    }
    return typeResult;
  }

  private CppTypeInfo mapType(final FType derived) {
    if (derived instanceof FStructType) {
      return createCustomTypeInfo(derived, STRUCT);
    } else if (derived instanceof FTypeDef) {
      return mapTypeDef((FTypeDef) derived);
    } else if (derived instanceof FEnumerationType) {
      return createEnumTypeInfo((FEnumerationType) derived);
    } else if (derived instanceof FArrayType) {
      CppTypeInfo innerType = mapType(((FArrayType) derived).getElementType());
      return CArrayMapper.createArrayReference(innerType);
    } else if (derived instanceof FMapType) {
      return mapMapType((FMapType) derived);
    } else {
      return new CppTypeInfo(VOID);
    }
  }

  private CppTypeInfo mapTypeDef(FTypeDef derived) {
    if (isInstanceId(derived)) {
      return createCustomTypeInfo((FInterface) derived.eContainer(), CLASS);
    } else {
      return mapType(derived.getActualType());
    }
  }

  private CppTypeInfo mapPredefined(final FTypeRef type) {
    FBasicTypeId typeId = type.getPredefined();
    switch (typeId.getValue()) {
      case FBasicTypeId.UNDEFINED_VALUE:
        return new CppTypeInfo(CType.VOID);
      case FBasicTypeId.INT8_VALUE:
        return new CppTypeInfo(CType.INT8);
      case FBasicTypeId.UINT8_VALUE:
        return new CppTypeInfo(CType.UINT8);
      case FBasicTypeId.INT16_VALUE:
        return new CppTypeInfo(CType.INT16);
      case FBasicTypeId.UINT16_VALUE:
        return new CppTypeInfo(CType.UINT16);
      case FBasicTypeId.INT32_VALUE:
        return new CppTypeInfo(CType.INT32);
      case FBasicTypeId.UINT32_VALUE:
        return new CppTypeInfo(CType.UINT32);
      case FBasicTypeId.INT64_VALUE:
        return new CppTypeInfo(CType.INT64);
      case FBasicTypeId.UINT64_VALUE:
        return new CppTypeInfo(CType.UINT64);
      case FBasicTypeId.BOOLEAN_VALUE:
        return new CppTypeInfo(CType.BOOL);
      case FBasicTypeId.FLOAT_VALUE:
        return new CppTypeInfo(CType.FLOAT);
      case FBasicTypeId.DOUBLE_VALUE:
        return new CppTypeInfo(CType.DOUBLE);
      case FBasicTypeId.STRING_VALUE:
        return CppTypeInfo.Companion.getSTRING();
      case FBasicTypeId.BYTE_BUFFER_VALUE:
        return byteBufferTypeInfo;
    }
    return new CppTypeInfo(VOID);
  }

  public CppTypeInfo createCustomTypeInfo(
      final FModelElement modelElement, final CppTypeInfo.TypeCategory category) {

    String baseApiName = cppNameResolver.getFullyQualifiedName(modelElement);
    String baseApiCall = CBridgeNameRules.getBaseApiCall(category, baseApiName);

    Include publicInclude = includeResolver.resolveInclude(modelElement);
    Include baseApiInclude = cppIncludeResolver.resolveInclude(modelElement);

    CType structCType = new CType(BASE_REF_NAME, publicInclude);

    return new CppTypeInfo(
        baseApiCall,
        structCType,
        structCType,
        category,
        Arrays.asList(
            publicInclude,
            baseApiInclude,
            BASE_HANDLE_IMPL_INCLUDE,
            CppLibraryIncludes.MEMORY,
            CppLibraryIncludes.NEW));
  }

  public static CppTypeInfo createNullableTypeInfo(
      final CppTypeInfo baseTypeInfo, final CppTypeRef cppTypeRef) {

    List<Include> includes = new LinkedList<>();
    includes.addAll(baseTypeInfo.getIncludes());
    includes.addAll(cppTypeRef.includes);

    CType structCType = new CType(BASE_REF_NAME);
    return new CppTypeInfo(
        cppTypeRef.fullyQualifiedName, structCType, structCType, CLASS, includes);
  }

  public CppTypeInfo createEnumTypeInfo(final FEnumerationType francaEnum) {

    Include publicInclude = includeResolver.resolveInclude(francaEnum);
    Include baseApiInclude = cppIncludeResolver.resolveInclude(francaEnum);

    CType enumCType = new CType(CBridgeNameRules.getEnumName(francaEnum), publicInclude);

    return new CppTypeInfo(
        cppNameResolver.getFullyQualifiedName(francaEnum),
        enumCType,
        enumCType,
        ENUM,
        Arrays.asList(publicInclude, baseApiInclude));
  }

  public CppTypeInfo createErrorTypeInfo(final FEnumerationType francaEnum) {
    CppTypeInfo errorEnumInfo = createEnumTypeInfo(francaEnum);
    errorEnumInfo.getFunctionReturnType().includes.add(CType.BOOL_INCLUDE);
    return errorEnumInfo;
  }

  private CppTypeInfo mapMapType(final FMapType francaMapType) {

    CType cType = new CType(BASE_REF_NAME);
    CppTypeInfo keyType = mapType(francaMapType.getKeyType());
    CppTypeInfo valueType = mapType(francaMapType.getValueType());

    List<Include> includes = new LinkedList<>();
    includes.add(Include.Companion.createInternalInclude(BASE_HANDLE_IMPL_FILE));
    includes.add(CppLibraryIncludes.MAP);

    String enumHash = null;
    if (keyType.getTypeCategory() == CppTypeInfo.TypeCategory.ENUM) {
      includes.add(CppLibraryIncludes.ENUM_HASH);
      enumHash = enumHashType;
    }

    return new CppMapTypeInfo(
        cppNameResolver.getFullyQualifiedName(francaMapType),
        cType,
        cType,
        includes,
        keyType,
        valueType,
        enumHash);
  }

  public String getEnumHashType() {
    return this.enumHashType;
  }
}

File: genium/src/main/java/com/here/genium/generator/cbridge/CBridgeTypeMapper.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.cbridge

import com.here.genium.generator.cbridge.CBridgeNameRules.BASE_HANDLE_IMPL_FILE
import com.here.genium.generator.cbridge.CBridgeNameRules.BASE_REF_NAME
import com.here.genium.generator.cpp.CppIncludeResolver
import com.here.genium.generator.cpp.CppLibraryIncludes
import com.here.genium.generator.cpp.CppNameResolver
import com.here.genium.generator.cpp.CppTypeMapper
import com.here.genium.model.cbridge.CArray
import com.here.genium.model.cbridge.CBridgeIncludeResolver
import com.here.genium.model.cbridge.CCollectionType
import com.here.genium.model.cbridge.CMap
import com.here.genium.model.cbridge.CSet
import com.here.genium.model.cbridge.CType
import com.here.genium.model.common.Include
import com.here.genium.model.cpp.CppTypeRef
import com.here.genium.model.lime.LimeList
import com.here.genium.model.lime.LimeBasicType
import com.here.genium.model.lime.LimeBasicType.TypeId
import com.here.genium.model.lime.LimeContainer
import com.here.genium.model.lime.LimeEnumeration
import com.here.genium.model.lime.LimeMap
import com.here.genium.model.lime.LimeNamedElement
import com.here.genium.model.lime.LimeSet
import com.here.genium.model.lime.LimeStruct
import com.here.genium.model.lime.LimeType
import com.here.genium.model.lime.LimeTypeDef
import com.here.genium.model.lime.LimeTypeHelper

class CBridgeTypeMapper(
    private val cppIncludeResolver: CppIncludeResolver,
    private val cppNameResolver: CppNameResolver,
    private val includeResolver: CBridgeIncludeResolver,
    private val internalNamespace: List<String>
) {
    private val genericsCollector = mutableMapOf<String, CCollectionType>()
    val generics: Map<String, CCollectionType> = genericsCollector

    private val byteBufferTypeInfo = CppTypeInfo(
        CppTypeMapper.BYTE_BUFFER_POINTER_TYPE.name,
        CType(BASE_REF_NAME),
        CType.BYTE_ARRAY_REF,
        CppTypeInfo.TypeCategory.BUILTIN_BYTEBUFFER,
        listOf(Include.createInternalInclude(BASE_HANDLE_IMPL_FILE))
    )

    fun mapType(limeType: LimeType): CppTypeInfo =
        when (limeType) {
            is LimeBasicType -> mapBasicType(limeType)
            is LimeTypeDef -> mapType(LimeTypeHelper.getActualType(limeType))
            is LimeContainer -> createCustomTypeInfo(limeType, CppTypeInfo.TypeCategory.CLASS)
            is LimeStruct -> createCustomTypeInfo(limeType, CppTypeInfo.TypeCategory.STRUCT)
            is LimeEnumeration -> createEnumTypeInfo(limeType)
            is LimeList -> {
                val result = createArrayTypeInfo(mapType(limeType.elementType.type))
                val arrayName = CBridgeNameResolver.getCollectionName(limeType)
                genericsCollector.putIfAbsent(arrayName, CArray(arrayName, result))
                result
            }
            is LimeMap -> {
                val keyType = mapType(limeType.keyType.type)
                val valueType = mapType(limeType.valueType.type)
                val hasStdHash = CppTypeMapper.hasStdHash(limeType.keyType)
                val result = createMapTypeInfo(keyType, valueType, hasStdHash)

                val mapName = CBridgeNameResolver.getCollectionName(limeType)
                val cMap = CMap(mapName, keyType, valueType, CppLibraryIncludes.MAP, hasStdHash)
                genericsCollector.putIfAbsent(mapName, cMap)
                result
            }
            is LimeSet -> {
                val elementType = mapType(limeType.elementType.type)
                val hasStdHash = CppTypeMapper.hasStdHash(limeType.elementType)
                val result = createSetTypeInfo(elementType, hasStdHash)

                val setName = CBridgeNameResolver.getCollectionName(limeType)
                val cSet = CSet(setName, elementType, CppLibraryIncludes.SET, hasStdHash)
                genericsCollector.putIfAbsent(setName, cSet)
                result
            }
            else -> CppTypeInfo(CType.VOID)
        }

    fun createCustomTypeInfo(
        limeElement: LimeNamedElement,
        category: CppTypeInfo.TypeCategory
    ): CppTypeInfo {
        val baseApiName = cppNameResolver.getFullyQualifiedName(limeElement)
        val baseApiCall = CBridgeNameRules.getBaseApiCall(category, baseApiName)

        val publicInclude = includeResolver.resolveInclude(limeElement)
        val structCType = CType(BASE_REF_NAME, publicInclude)

        val includes = listOf(publicInclude, CppLibraryIncludes.OPTIONAL) +
            cppIncludeResolver.resolveIncludes(limeElement) + BASE_HANDLE_IMPL_INCLUDE +
            CppLibraryIncludes.MEMORY + CppLibraryIncludes.NEW

        return CppTypeInfo(baseApiCall, structCType, structCType, category, includes)
    }

    fun createEnumTypeInfo(limeType: LimeType, asErrorType: Boolean = false): CppTypeInfo {
        val publicInclude = includeResolver.resolveInclude(limeType)
        val baseApiIncludes = cppIncludeResolver.resolveIncludes(limeType)

        val typeName = CBridgeNameRules.getTypeName(limeType)
        val functionReturnTypeIncludes = listOf(publicInclude) +
            if (asErrorType) listOf(CType.BOOL_INCLUDE) else emptyList()

        return CppTypeInfo(
            name = cppNameResolver.getFullyQualifiedName(limeType),
            cType = CType(typeName, publicInclude),
            functionReturnType = CType(typeName, functionReturnTypeIncludes),
            typeCategory = CppTypeInfo.TypeCategory.ENUM,
            includes = listOf(publicInclude) + baseApiIncludes
        )
    }

    private fun mapBasicType(limeType: LimeBasicType) =
        when (limeType.typeId) {
            TypeId.VOID -> CppTypeInfo(CType.VOID)
            TypeId.INT8 -> CppTypeInfo(CType.INT8)
            TypeId.UINT8 -> CppTypeInfo(CType.UINT8)
            TypeId.INT16 -> CppTypeInfo(CType.INT16)
            TypeId.UINT16 -> CppTypeInfo(CType.UINT16)
            TypeId.INT32 -> CppTypeInfo(CType.INT32)
            TypeId.UINT32 -> CppTypeInfo(CType.UINT32)
            TypeId.INT64 -> CppTypeInfo(CType.INT64)
            TypeId.UINT64 -> CppTypeInfo(CType.UINT64)
            TypeId.BOOLEAN -> CppTypeInfo(CType.BOOL)
            TypeId.FLOAT -> CppTypeInfo(CType.FLOAT)
            TypeId.DOUBLE -> CppTypeInfo(CType.DOUBLE)
            TypeId.STRING -> CppTypeInfo.STRING
            TypeId.BLOB -> byteBufferTypeInfo
            TypeId.DATE -> CppTypeInfo.DATE
        }

    private fun createMapTypeInfo(
        keyType: CppTypeInfo,
        valueType: CppTypeInfo,
        hasStdHash: Boolean
    ): CppMapTypeInfo {
        val keyTypeName = keyType.name
        val cppName = if (hasStdHash) {
            "std::unordered_map<$keyTypeName, ${valueType.name}>"
        } else {
            val namespace = internalNamespace.joinToString("::")
            "std::unordered_map<$keyTypeName, ${valueType.name}, $namespace::hash<$keyTypeName>>"
        }
        return CppMapTypeInfo(
            cppName,
            CType(BASE_REF_NAME),
            listOf(
                Include.createInternalInclude(BASE_HANDLE_IMPL_FILE),
                CppLibraryIncludes.MAP
            ),
            keyType,
            valueType
        )
    }

    private fun createSetTypeInfo(elementType: CppTypeInfo, hasStdHash: Boolean): CppSetTypeInfo {
        val keyTypeName = elementType.name
        val cppName = if (hasStdHash) {
            "std::unordered_set<$keyTypeName>"
        } else {
            val namespace = internalNamespace.joinToString("::")
            "std::unordered_set<$keyTypeName, $namespace::hash<$keyTypeName>>"
        }
        return CppSetTypeInfo(
            cppName,
            CType(BASE_REF_NAME),
            listOf(
                Include.createInternalInclude(BASE_HANDLE_IMPL_FILE),
                CppLibraryIncludes.SET
            ),
            elementType
        )
    }

    companion object {
        val BASE_HANDLE_IMPL_INCLUDE = Include.createInternalInclude(BASE_HANDLE_IMPL_FILE)

        private fun createArrayTypeInfo(elementTypeInfo: CppTypeInfo) =
            CppArrayTypeInfo(
                "std::vector<${elementTypeInfo.name}>",
                CType(BASE_REF_NAME),
                CType(BASE_REF_NAME),
                listOf(
                    Include.createInternalInclude(BASE_HANDLE_IMPL_FILE),
                    CppLibraryIncludes.VECTOR
                ),
                elementTypeInfo
            )

        fun createNullableTypeInfo(
            baseTypeInfo: CppTypeInfo,
            cppTypeRef: CppTypeRef
        ) = CppTypeInfo(
            cppTypeRef.fullyQualifiedName,
            CType(BASE_REF_NAME),
            CType(BASE_REF_NAME),
            CppTypeInfo.TypeCategory.CLASS,
            baseTypeInfo.includes + cppTypeRef.includes + CppLibraryIncludes.OPTIONAL
        )
    }
}
